### PR TITLE
Move FeeConsensus into modules

### PR DESCRIPTION
**Pull request by rndhouse**
Addresses issue: https://github.com/fedimint/minimint/issues/313

### CODE REVIEW

**Comment by rndhouse** [comment]
One of: [@elsirion, @justinmoon,  @jkitman] please. 

Could I have an opinion on commit https://github.com/fedimint/minimint/pull/325/commits/25b2cf7ce4ccd9bf5ce4530dacc4b275a426cc3c please? Does that sort of change seem reasonable to you? I'm not sure if `minimint_wallet::PegOut` should know about consensus fees.

As far as I can tell, `FeeConsensus` fields are only ever set to `minimint_api::Amount::ZERO`. Will they be configurable in the future? Should these fields be part of the config structures (e.g. `minimint_wallet::config::WalletConfig`)? If they should be in configs then I'm not sure how to get `ServerConfig` into `minimint_core::transaction::Output::fee` without moving it to `minimint_core`.

`InputMeta` was mentioned as candidate housing. But it isn't on the ln/wallet/mint module level. Could you expand on that idea a bit please? And also for `OutputMeta` I suppose.

**Comment by justinmoon** [comment]
Thanks for getting this started!

> FeeConsensus fields are only ever set to minimint_api::Amount::ZERO. Will they be configurable in the future?

Yes they will be configurable in the future, but aren't configurable at all now. Someday I think a live federation should be able to alter fee-rates while the federation is running if sufficient number of servers agree to it. 

After thinking about it a little more, I think there are 2 separate things happening here and could be implemented separately in 2 successive PRs:
(1) Move fee information into modules
(2) Add consensus logic so that fee rates can be adjusted dynamically

For (1), I think each module could have their own fee-related struct defined in that module's crate. Then each of these structs could be placed in both the client and server configs. Taking wallet module as an example, server-side part could go in `WalletConfig` and client-side part could go in `WalletClientConfig`. Maybe try this with the wallet module (`fee_peg_out_abs` and `fee_peg_in_abs`) to get started?

I'm not exactly sure how we will do (2).

**Comment by rndhouse** [comment]
Thank you for the reply. I've moved the `FeeConsensus` structs into the different modules configs.

One drawback of my approach is that I've added a method to `ServerConfig` and `ClientConfig`:
```rust
    pub fn fee_consensus(&self) -> minimint_core::config::FeeConsensus {
        minimint_core::config::FeeConsensus {
            wallet: self.wallet.fee_consensus.clone(),
            mint: self.mint.fee_consensus.clone(),
            ln: self.ln.fee_consensus.clone(),
        }
    }
```

This avoids the need to move different configs into `minimint_core::transaction::Input::fee` and `minimint_core::transaction::Output::fee`. Which would probably require moving `ServerConfig` (and some `minimint::net` components into `minimint_core`).

I think this PR completes the issue, but let me know what you think my current changes.

**Review by elsirion — CHANGES_REQUESTED** [review]
This seems like a good first step towards #313. Ideally we could get rid of `FeeConsensus` completely. You already split it up and it feels rather redundant now. This will require more refactoring around the transaction validation logic though.

The goal should probably be to get rid of the `TransactionItem` trait and do amount validation using `apply_input`/`apply_output` which already return the amounts. This will also simplify further modularization.

File: minimint-api/src/lib.rs [hunk 1]
```diff
@@ -129,6 +129,12 @@ impl Amount {
     }
 }
 
+impl Default for Amount {
+    fn default() -> Self {
+        Amount::ZERO
+    }
+}
+
```
[the remark this answers is not in the document]
Comment (justinmoon) [reply minimint-api/src/lib.rs:137]: I think he means that `Amount` is used for many different things now, beyond just fees. A zero default might make sense for fees, but not for other things like a transaction output. So let's remove this `Default` impl on `Amount`.

File: minimint/src/config.rs [hunk 2]
```diff
@@ -96,15 +93,6 @@ impl GenerateConfig for ServerConfig {
         let (ln_server_cfg, ln_client_cfg) =
             LightningModuleConfig::trusted_dealer_gen(peers, max_evil, &(), &mut rng);
 
-        let fee_consensus = FeeConsensus {
-            fee_coin_spend_abs: minimint_api::Amount::ZERO,
-            fee_peg_in_abs: minimint_api::Amount::ZERO,
-            fee_coin_issuance_abs: minimint_api::Amount::ZERO,
```
Comment (justinmoon) [inline minimint/src/config.rs:102]: I vote that we keep `fee_coin_issuance_abs`. Issuing a blind signature takes computational power so I can imagine why a federation might choose to charge a fee for it.

**Comment by justinmoon** [comment]
I wonder if it makes sense to have a `Fee` type with a default of zero, which impls `Into<Amount>`. Deriving `Default` for `FeeConsensus` is nice, but having a default of zero for transaction output amounts (for instance) doesn't seem right ...

File: modules/minimint-mint/src/config.rs [hunk 3]
```diff
@@ -65,8 +68,24 @@ impl GenerateConfig for MintConfig {
...
+            fee_consensus: FeeConsensus::default(),
         };
 
         (mint_cfg, client_cfg)
     }
 }
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct FeeConsensus {
+    pub coin_issuance_abs: minimint_api::Amount,
```
Comment (rndhouse) [inline modules/minimint-mint/src/config.rs:80]: Not certain if this is the right place for this (unused) field. Please double check.

File: minimint-api/src/lib.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (rndhouse) [reply minimint-api/src/lib.rs]: Thank you both for taking a look!

That does make sense to me. I've removed `Default` impl on `Amount` in the latest rebase.

File: minimint/src/config.rs (hunk 2 above)
Comment (rndhouse) [reply minimint/src/config.rs:102]: I've retained this field in the latest rebase. I've put the field in the mint module's config. Next to `coin_spend_abs`. Not sure if that's the right spot, please confirm.

File: modules/minimint-mint/src/config.rs (hunk 3 above)
Comment (justinmoon) [reply modules/minimint-mint/src/config.rs:80]: Yup that looks right.

**Comment by rndhouse** [comment]
> I wonder if it makes sense to have a `Fee` type with a default of zero, which impls `Into<Amount>`. Deriving `Default` for `FeeConsensus` is nice, but having a default of zero for transaction output amounts (for instance) doesn't seem right ...

A `Fee` type with a default of zero sounds right to me. I would probably make that change in another PR, but I don't mind either way.

**Comment by rndhouse** [comment]
> This seems like a good first step towards #313. Ideally we could get rid of `FeeConsensus` completely. You already split it up and it feels rather redundant now. This will require more refactoring around the transaction validation logic though.
> 
> The goal should probably be to get rid of the `TransactionItem` trait and do amount validation using `apply_input`/`apply_output` which already return the amounts. This will also simplify further modularization.

@elsirion That makes sense to me on a high level. I'll reach out to you on discord for a bit more hints if you don't mind!

**Comment by rndhouse** [comment]
@elsirion rebase complete!

File: client/client-lib/src/lib.rs [hunk 4]
```diff
@@ -218,7 +217,12 @@ impl<T: AsRef<ClientConfig> + Clone> Client<T> {
     ) -> Result<TransactionId> {
         Ok(self
             .mint_client()
-            .submit_tx_with_change(&self.context.config.as_ref().fee_consensus, tx, batch, rng)
+            .submit_tx_with_change(
+                &self.context.config.as_ref().fee_consensus(),
```
Comment (elsirion) [inline client/client-lib/src/lib.rs:221]: nit: I think the `as_ref` isn't needed here. Function calls do some automatic coercion.

**Review by elsirion — APPROVED** [review]
Great first step to fix https://github.com/fedimint/minimint/issues/313! As discussed the remainder will be in a separate PR. LGTM.

**Pull request merged**